### PR TITLE
feat(side-dag): implement side_dag CLI command

### PR DESCRIPTION
**Pull request by glevco**
Depends on https://github.com/HathorNetwork/hathor-core/pull/1040

### Motivation

Implement a new `side_dag` CLI command that is able to run two independent full nodes, each in its own separate process. The command accepts the same options as the `run_node` command, however options can be duplicate with a `--side-dag` prefix. Each full node is configured with the respective set of options. Processes are automatically terminated when one of the full nodes exits.

Here's an example on how this command could be used:

 ```bash
  $ python -m hathor side_dag
      --testnet
      --procname-prefix testnet-
      --memory-storage
      --disable-logs
      --side-dag-config-yaml ./my-side-dag.yml
      --side-dag-procname-prefix my-side-dag-
      --side-dag-memory-storage
      --side-dag-json-logs
```

### Acceptance Criteria

- Implement new `cli.util.process_logging_options()`.
- Change `execute_from_command_line()` so it supports modules opting out of the `setup_logging()` call.
- Update `create_parser()` so it accepts a custom env var prefix.
- Add support to disabling logs.
- Implement `side_dag` CLI command.
- Implement `HathorSettings.GENESIS_TOKENS` calculated property, removing the `GENESIS_TOKENS` constant.

### Checklist

- [x] If you are requesting a merge into `master`, confirm this code is production-ready and can be included in future releases as soon as it gets merged

### CODE REVIEW

**Review by msbrogli — CHANGES_REQUESTED** [review]

File: hathor/cli/side_dag.py [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+class HathorProcessTerminated:
+    pass
+
+
+@dataclass(frozen=True, slots=True)
+class SideDagProcessTerminated:
+    pass
+
+
+class SideDag(RunNode):
```
Comment (msbrogli) [inline hathor/cli/side_dag.py:61]: Maybe rename to `SideDagRunNode`?

File: hathor/cli/side_dag.py [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+            argv_list = side_dag_argv
+
+        argv_list.append(arg)
+        if value is not None:
+            argv_list.append(value)
+
+    return hathor_node_argv, side_dag_argv
+
+
+def _run_side_dag_node(argv: list[str], *, hathor_node_process: Process, conn: 'Connection') -> None:
```
Comment (msbrogli) [inline hathor/cli/side_dag.py:131]: If I got it right, the current process will spawn a new process for the mainnet full node and run the side-dag full node in the current process. Is that right? Isn't it better to spawn two new processes, one for each full node and develop a process monitor that gracefully terminates when one of the processes dies or freezes?

About signals, all processes would have to handle signals.

**Review by msbrogli — CHANGES_REQUESTED** [review]

File: hathor/cli/util.py [hunk 3]
```diff
@@ -230,7 +271,12 @@ def setup_logging(
 
     def kwargs_formatter(_, __, event_dict):
         if event_dict and event_dict.get('event') and isinstance(event_dict['event'], str):
-            event_dict['event'] = event_dict['event'].format(**event_dict)
+            try:
```
[the remark this answers is not in the document]
Comment (msbrogli) [reply hathor/cli/util.py]: Ok. Thanks!

File: hathor/cli/util.py [hunk 4]
```diff
@@ -275,10 +321,9 @@ def twisted_structlog_observer(event):
             if failure is not None:
                 kwargs['exc_info'] = (failure.type, failure.value, failure.getTracebackObject())
             twisted_logger.log(level, msg, **kwargs)
-        except Exception as e:
-            print('error when logging event', e)
-            for k, v in event.items():
-                print(k, v)
+        except Exception:
+            logger = structlog.get_logger()
+            logger.exception('error when logging event')
```
[the remark this answers is not in the document]
Comment (msbrogli) [reply hathor/cli/util.py]: Ok. Thanks!

File: hathor/cli/side_dag.py [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+    hathor_node_process = _start_hathor_node_process(hathor_node_argv, capture_stdout=capture_stdout, conn=conn1)
+
+    debug, logging_output, sentry = process_logging_options(side_dag_argv)
+    setup_logging(debug=debug, capture_stdout=capture_stdout, logging_output=logging_output, sentry=sentry)
+    logger.info('starting nodes', hathor_node_pid=hathor_node_process.pid, side_dag_pid=os.getpid())
+
+    _run_side_dag_node(side_dag_argv, hathor_node_process=hathor_node_process, conn=conn2)
+
+
+def _process_argv(argv: list[str]) -> tuple[list[str], list[str]]:
```
[the remark this answers is not in the document]
Comment (msbrogli) [reply hathor/cli/side_dag.py]: Thanks!

File: hathor/cli/side_dag.py (hunk 1 above)
Comment (msbrogli) [reply hathor/cli/side_dag.py]: Thanks!

File: hathor/cli/side_dag.py (hunk 2 above)
Comment (msbrogli) [reply hathor/cli/side_dag.py:136]: As discussed in person, let's do a monitor process in a separate PR so we can skip it for now.

File: hathor/conf/settings.py [hunk 6]
```diff
@@ -55,7 +54,9 @@ class HathorSettings(NamedTuple):
     # Genesis pre-mined tokens
     GENESIS_TOKEN_UNITS: int = GENESIS_TOKEN_UNITS
 
-    GENESIS_TOKENS: int = GENESIS_TOKENS
+    @property
+    def GENESIS_TOKENS(self) -> int:
```
[the remark this answers is not in the document]
Comment (msbrogli) [reply hathor/conf/settings.py:58]: I guess we can discuss it separately. We can add a consistency check before initializing a full node, for instance.

File: hathor/cli/side_dag.py (hunk 2 above)
Comment (glevco) [reply hathor/cli/side_dag.py:136]: Created an issue for this: https://github.com/HathorNetwork/hathor-core/issues/1051

File: hathor/conf/settings.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (glevco) [reply hathor/conf/settings.py]: Sure, removed it in https://github.com/HathorNetwork/hathor-core/pull/1039/commits/fbf9caff99ec80ab1786bb2e748f1b4e6bdb8e9e

[the remark this answers is not in the document]
Comment (msbrogli) [reply hathor/conf/settings.py]: Thanks!

**Review by msbrogli — APPROVED** [review]

File: hathor/cli/side_dag.py [hunk 7]
```diff
@@ -0,0 +1,220 @@
...
+    ```bash
+    $ python -m hathor side_dag
+        --testnet
+        --procname-prefix testnet-
+        --memory-storage
+        --disable-logs
+        --side-dag-config-yaml ./my-side-dag.yml
+        --side-dag-procname-prefix my-side-dag-
+        --side-dag-memory-storage
+        --side-dag-json-logs
```
Comment (jansegre) [inline hathor/cli/side_dag.py:85]: Does `--side-dag-json-logs` need to be a separate option from `--json-logs`? The reason `--json-logs` exist is to make all the logs that the stdout/stderr (as observed by whoever started the process) be written as `JSON` entries (one per line), instead of formatted entries (which can be colored). The option exists independently of which subcommand is being used even for those not based on `RunNode`.

What do you think about using only one argument instead?

File: hathor/cli/side_dag.py [hunk 8]
```diff
@@ -0,0 +1,220 @@
...
+    hathor_node_process.start()
+    return hathor_node_process
+
+
+def _run_hathor_node(argv: list[str], run_node_cmd: type[RunNode], capture_stdout: bool, conn: 'Connection') -> None:
+    """Function to be called by a background process to run the Hathor full node."""
+    from hathor.cli.util import process_logging_options, setup_logging
+
+    # We don't terminate via SIGINT directly, instead the side-dag process will terminate us.
+    signal.signal(signal.SIGINT, lambda _, __: None)
```
Comment (jansegre) [inline hathor/cli/side_dag.py:191]: Will this suppress the normal behavior of raising a `KeyboardInterrupt` when the process receives a `SIGINT`?

File: hathor/cli/side_dag.py [hunk 9]
```diff
@@ -0,0 +1,220 @@
...
+        message = conn.recv()
+        assert isinstance(message, SideDagProcessTerminated)
+        logger.critical('hathor node terminated because side-dag process was terminated')
+        return
+
+    conn.send(HathorProcessTerminated())
+    logger.critical('terminating side-dag node...')
+    parent_pid = os.getppid()
+    parent_process = psutil.Process(parent_pid)
+    parent_process.terminate()
```
Comment (jansegre) [inline hathor/cli/side_dag.py:220]: What does `terminate` do? Does it just send a `SIGTERM`?

File: hathor/cli/side_dag.py (hunk 7 above)
Comment (glevco) [reply hathor/cli/side_dag.py:85]: It could be a single option, the reason I implemented it separately is to be able to configure each full node's logging independently. I created the `LoggingOutput` enum and those arguments simply select one of the enum options. It accepts either `--[side-dag-]disable-logs`, `--[side-dag-]json-logs`, or not set, which respectively map to `NULL`, `JSON`, or `PRETTY` (the default).

So for example, if I'm running locally, I would probably set only `--disable-logs`, which would disable any logs from the hathor node, and the side-dag logs would be pretty outputted to stdout/sderr.

If I'm running on a server, I would probably set `--json-logs --side-dag-json-logs`, to get json outputs from both nodes. But one could easily want to discard hathor logs, for example, in which case they would set `--disable-logs --side-dag-json-logs`.

In summary, this is just giving more flexibility. Do you prefer to keep a single option?

File: hathor/cli/side_dag.py (hunk 8 above)
Comment (glevco) [reply hathor/cli/side_dag.py:191]: Yes!

File: hathor/cli/side_dag.py (hunk 9 above)
Comment (glevco) [reply hathor/cli/side_dag.py:220]: > What does `terminate` do? Does it just send a `SIGTERM`? Because that alone might not be enough to reliably terminate the parent process

Yes, it just sends a `SIGTERM`. Why not?

> there's an issue about doing a monitor process in a separate PR, that's fine by me.

Yep, it will be a future improvement.

Comment (glevco) [reply hathor/cli/side_dag.py:220]: I added this in the issue: https://github.com/HathorNetwork/hathor-core/issues/1051#issuecomment-2153518597

File: hathor/cli/side_dag.py (hunk 7 above)
Comment (jansegre) [reply hathor/cli/side_dag.py:85]: I see, I'm not against the flexibility, but I'm not a fan of having the behavior of `--json-logs` sort of change (from having all logs in JSON, to some of the logs in JSON), I understand that production would just need to use `--json-logs --side-dag-json-logs` but I think that's a bit counter-intuitive.

I'd rather have the `side-dag` process use the same behavior (`NULL`, `JSON`, `PRETTY`) as the main log, unless there's an argument that changes it.

What about making the options `--json-logs` and `--disable-logs` take an optional argument: `both`, `main`, `side-dag`? And the default when the argument is present would be `both`. Since these two are special options that are processed before the `RunNode` ones, I think it would be an acceptable compromise.

Comment (jansegre) [reply hathor/cli/side_dag.py:85]: This can be implemented with [ArgumentParser.parse_known_args](https://docs.python.org/3/library/argparse.html#argparse.ArgumentParser.parse_known_args) instead of doing it manually like is was before.

**Review by jansegre — APPROVED** [review]

File: hathor/cli/side_dag.py (hunk 7 above)
Comment (glevco) [reply hathor/cli/side_dag.py:93]: Done in https://github.com/HathorNetwork/hathor-core/pull/1039/commits/90a0606cb53cc1939157c4b172cf5ad8d380e6f0

**Review by jansegre — APPROVED** [review]

File: hathor/cli/side_dag.py (hunk 9 above)
Comment (msbrogli) [reply hathor/cli/side_dag.py:277]: @jansegre I had a similar conversation about this topic with @glevco during my review.

**Review by msbrogli — APPROVED** [review]
